### PR TITLE
[Dev] Introduce forward sharing state and enable repeated MTP sharing

### PR DESCRIPTION
**Pull request by buptzyb**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?

Introduce a shared infrastructure for cooperating components to exchange per-forward payloads, migrate existing DSA cross-layer IndexShare onto it, and use it to implement independently selectable index and latent-KV sharing across repeated MTP prediction depths.

## Implementation and review order

The PR is organized into three commits:

1. **Generic forward sharing state.** Introduce `ForwardSharingState`, a type-keyed payload registry with explicit lifetime management. Payloads define their own tensor layout and per-layer indexing. State is carried through existing packed-sequence parameters, attention masks, or config, without adding a new Transformer-layer forward argument. Outer scopes clear payloads on exit; nested scopes join the active lifetime. Temporary state binding and per-checkpoint snapshots preserve the sharing records needed by existing checkpoint/schedule paths. Ordinary DSA cross-layer IndexShare is migrated to this infrastructure.
2. **Repeated MTP index sharing.** Reuse depth-0 sparse-attention indices at later depths of the same repeated physical MTP layer. Ordinary IndexShare still determines how depth 0 obtains its indices; repeated-MTP sharing controls their reuse at subsequent depths. DSA owns the tensor payload, while MTP controls source/consumer sequencing.
3. **Repeated MTP latent-KV sharing.** Independently reuse depth-0 post-RoPE, post-TP/CP-layout latent KV, avoiding repeated KV computation and CP gathering. Shared KV stays connected to autograd so consumer gradients accumulate into the source projection.

The common infrastructure manages payload lookup, binding, and cleanup; it does not prescribe DSA tensor layouts or make arbitrary sharing/recompute combinations safe.

## Feature scope and compatibility

Set `mtp_use_repeated_layer: true`, `mtp_num_layers > 1`, and select `"sparse_attention_index"`, `"latent_kv"`, or both through `mtp_repeated_layer_shared_components`. Repeated-MTP sharing currently supports the DSA GPT MTP path only. Queries and sparse attention remain query-dependent and are computed at every depth. This is training-time activation reuse, not inference KV caching.

- Existing selective recompute of `mlp`, `moe`, `moe_act`, `shared_experts`, `layernorm`, and `mhc` remains supported, subject to those modules' existing requirements.
- `mla_up_proj` recompute supports index-only sharing, but not latent-KV sharing in this PR.
- Full recompute and selective `core_attn` recompute remain unsupported with repeated-MTP sharing. Set `recompute_modules` explicitly, since selective recompute defaults to `core_attn`.
- Attention CUDA graph capture remains unsupported with repeated-MTP sharing; MoE-only scopes remain compatible. Unsupported inference and hybrid-layer combinations are rejected explicitly.

The dependent draft #7107 extends repeated-MTP recompute compatibility; it is not required for the ordinary IndexShare checkpoint handling or the selective combinations supported here.

## Issue tracking

Related to #6392.

## Testing

- Focused local CPU regression suite at `90a3a3fdb`: **172 passed**, with 6 GPU-marked cases deselected.
- Coverage includes state binding and cleanup, nested lifetimes, checkpoint snapshots, ordinary IndexShare, independently selected repeated-MTP components, and compatibility validation.
- Additional tests cover CP layouts, forward/gradient parity, TP/SP layout handling, one-time KV/CP gathering, and PP/VPP placement. GPU execution is not implied by the local CPU result above.
- `git diff --check`, Black, isort, pylint, and ruff pass. The formatter's advisory mypy step still reports missing dependencies and type errors; this is not a clean type-check result.
- Results from the separate recompute follow-up are not included in this PR's test counts.

### Pre-checks

- [x] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [x] I have added proper typing to my code
- [x] I have added relevant documentation
- [x] I have run the autoformatter on this PR

## Main PR

Target-main counterpart: [NVIDIA/Megatron-LM#6874](https://github.com/NVIDIA/Megatron-LM/pull/6874).

### CODE REVIEW

File: megatron/core/transformer/experimental_attention_variant/absorbed_mla.py [hunk 1]
```diff
@@ -851,6 +894,9 @@ def custom_forward(*inputs):
             up_v_weight = inputs[5]
             attn_mask_type = inputs[6]
             attn_mask_type = AttnMaskType(attn_mask_type.item())
+            core_attention_kwargs = {}
+            if mtp_dsa_context is not None:
+                core_attention_kwargs["mtp_dsa_context"] = mtp_dsa_context
```
[the remark this answers is not in the document]
Comment (buptzyb) [reply megatron/core/transformer/experimental_attention_variant/absorbed_mla.py:893]: Fixed in `d8300ca38`. `TransformerConfig` now rejects `dsa_mtp_index_kv_share` with selective recompute whenever `core_attn` is present, including after the default recompute module list is materialized. The error directs users to selective `mla_up_proj`, full recompute, or disabling sharing. `test_iteration_sharing_rejects_default_selective_core_attention_recompute` covers the rejection, and the adjacent non-sharing test confirms the existing default remains valid when sharing is off. The final GB200 run passed all 37 tests in this file.

File: megatron/core/transformer/experimental_attention_variant/absorbed_mla.py [hunk 2]
```diff
@@ -562,23 +576,36 @@ def get_query_key_value_tensors(
...
+            """Select the RoPE rows used by both the query and key branches."""
+            if inference_context is not None:
+                sequence_start = inference_context.sequence_len_offset
+                sequence_end = sequence_start + num_tokens
+                return rotary_pos_emb[sequence_start:sequence_end]
+            if not thd_packed_seq or self.config.context_parallel_size == 1:
+                # Packed CP keeps the full embedding so every local segment can address its
+                # original positions. Other layouts use the local query length.
+                return rotary_pos_emb[:num_tokens]
+            return rotary_pos_emb
```
[the remark this answers is not in the document]
Comment (buptzyb) [reply megatron/core/transformer/experimental_attention_variant/absorbed_mla.py:602]: Confirmed that the non-sharing RoPE behavior is unchanged. Previously `q_len = q.size(0)` selected the same coordinate window for both Q and K. After factorization, Q uses `q.size(0)` and K receives `q_absorbed.size(0)`; the code asserts `q_absorbed.shape[:-1] == q.shape[:-1]`, so those lengths are identical. The non-sharing selective-recompute regression pins the original joint Q/K checkpoint path, and the TP2/SP test covers the distributed row layout. Both passed in the 37-test GB200 run.

File: megatron/core/transformer/experimental_attention_variant/absorbed_mla.py [hunk 3]
```diff
@@ -714,15 +738,33 @@ def qkv_up_proj_and_rope_apply(q_compressed, kv_compressed, k_pos_emb, rotary_po
...
+                else:
+                    kv_compressed = shared_key
+            else:
+                q_absorbed, kv_compressed = self.qkv_up_checkpoint.checkpoint(
+                    qkv_up_proj_and_rope_apply,
+                    q_compressed,
+                    kv_compressed,
+                    k_pos_emb,
+                    rotary_pos_emb,
+                )
```
[the remark this answers is not in the document]
Comment (buptzyb) [reply megatron/core/transformer/experimental_attention_variant/absorbed_mla.py:752]: This condition intentionally includes iteration 0. The source K must remain live after source attention so later MTP iterations can consume it. If iteration 0 used the joint `CheckpointWithoutOutput` path, K would be a checkpoint output and `discard_output_and_register_recompute()` would resize its storage before consumers run. Therefore the source iteration checkpoints Q only and builds K outside that checkpoint. The native parity test verifies retained source storage, one KV/indexer execution, and output/input/parameter-gradient parity across packed/unpacked plus fused/unfused RoPE cases; all variants passed on GB200.

File: megatron/core/transformer/multi_token_prediction.py [hunk 4]
```diff
@@ -2520,10 +2611,20 @@ def checkpoint_handler():
...
+        if mtp_dsa_context is not None and mtp_dsa_context.is_source:
+            if isinstance(outputs, torch.Tensor):
+                return outputs, mtp_dsa_context.require_source_tensors()
+            hidden_states, shared_key, shared_topk_indices, shared_topk_length = outputs
+            return hidden_states, MTPDSASharedTensors(
+                shared_key, shared_topk_indices, shared_topk_length
+            )
+        if mtp_dsa_context is not None:
+            return outputs, mtp_dsa_context.shared_tensors
         return outputs
```
[the remark this answers is not in the document]
Comment (buptzyb) [reply megatron/core/transformer/multi_token_prediction.py:2634]: Fixed in `d8300ca38`. `_checkpointed_forward` now tracks `used_checkpoint` explicitly instead of inferring provenance from the output type. Uniform recompute unpacks graph-connected checkpoint outputs into the context. The warned block fallback performs direct forward, validates that the caller-owned context captured source tensors, and returns the normal tensor result. `test_recompute_path_threads_shared_kv_through_iteration_context` covers native uniform, TE/MXFP8 uniform, and block fallback including source/consumer and parameter-gradient parity; it passed on GB200.

File: megatron/core/transformer/multi_token_prediction.py [hunk 5]
```diff
@@ -2619,9 +2731,18 @@ def forward(
...
+            if mtp_dsa_context is not None and mtp_dsa_context.is_source:
+                shared_tensors = mtp_dsa_context.require_source_tensors()
+            elif mtp_dsa_context is not None:
+                shared_tensors = mtp_dsa_context.shared_tensors
+            else:
+                shared_tensors = None
 
         self.cp_group = _orig_cp_group
+        if mtp_dsa_context is not None:
+            return hidden_states, input_ids, position_ids, padding_mask, shared_tensors
```
[the remark this answers is not in the document]
Comment (buptzyb) [reply megatron/core/transformer/multi_token_prediction.py]: Fixed in `d8300ca38`. `MultiTokenPredictionLayer.forward` always preserves the legacy four-element return schema; shared tensors now travel through `MTPDSAIterationContext`. `_checkpointed_forward` likewise returns only hidden states and publishes graph-connected source tensors through that context. `MultiTokenPredictionBlock` passes the new keyword only when sharing is enabled, preserving fixed-signature custom layers while the feature is default-off. The dedicated legacy-signature regression and recompute/source-threading tests passed on GB200.

File: megatron/core/transformer/experimental_attention_variant/dsa.py [hunk 6]
```diff
@@ -2016,6 +2034,11 @@ def _build_kv_reorder_idx(local_len):
 
         skv = key.size(0)
 
+        if mtp_dsa_context is not None and mtp_dsa_context.is_source:
+            source_key = key
+        else:
+            source_key = None
```
[the remark this answers is not in the document]
Comment (buptzyb) [reply megatron/core/transformer/experimental_attention_variant/dsa.py:2311]: Confirmed. In the exercised Absorbed MLA TP2/SP path, DSA receives global query rows. `test_tp2_sequence_parallel_reuses_global_source_kv_and_topk` pins both the global key length and `shared.topk_indices.size(1) == global_tokens`, then verifies backward flow into source and consumer inputs. For the generic sequence-local branch, the local row window depends only on TP rank and local row count; repeated iterations remain on the same rank and microbatch, so the window is iteration-invariant. Consumers reuse the exact source tensor and do not slice it again. The TP2/SP test passed on GB200.

File: megatron/core/transformer/experimental_attention_variant/dsa.py [hunk 7]
```diff
@@ -2040,7 +2063,7 @@ def _build_kv_reorder_idx(local_len):
         qr = qr.detach()
 
         indexer_loss_coeff = self.config.dsa_indexer_loss_coeff or 0.0
-        computes_topk = not self.skip_topk
+        computes_topk = not self.skip_topk and not reuse_mtp_source
         use_indexer_loss = (
             self.training and torch.is_grad_enabled() and indexer_loss_coeff > 0 and computes_topk
         )
```
[the remark this answers is not in the document]
Comment (buptzyb) [reply megatron/core/transformer/experimental_attention_variant/dsa.py:2340]: This is intentional and now documented. Sharing executes one indexer/top-k computation, so its auxiliary loss is computed and tracked once; multiplying it by `mtp_num_layers` would charge the same indexer result repeatedly. Recipes that require a particular total contribution should retune `dsa_indexer_loss_coeff`. `test_repeated_mtp_indexer_loss_charge_count` covers sharing/non-sharing with a nonzero coefficient, and `test_source_indexer_loss_and_gradients_match_unshared_oracle` verifies that source loss plus indexer parameter gradients match one unshared invocation. Both passed on GB200.

File: megatron/core/transformer/transformer_config.py [hunk 8]
```diff
@@ -3190,6 +3203,17 @@ def _scope_to_str(s):
...
+            captures_attention = (
+                self.cuda_graph_impl == "full_iteration"
+                or not self.cuda_graph_modules
+                or CudaGraphModule.attn in self.cuda_graph_modules
+            )
+            if captures_attention:
+                raise ValueError(
+                    "dsa_mtp_index_kv_share does not yet support CUDA graph scopes "
+                    "that capture attention. MoE-only CUDA graph scopes remain supported."
+                )
```
[the remark this answers is not in the document]
Comment (buptzyb) [reply megatron/core/transformer/transformer_config.py]: Fixed in `30d32e4a5`. Sharing validation now reuses the existing `cuda_graph_captures_attention` predicate. The rejection is also scoped to per-layer graph implementations: `full_iteration` remains allowed because it contains the complete MTP producer-consumer chain. Tests cover attention-capturing local/TE scopes and acceptance of MoE-only plus full-iteration scopes; they passed in the final GB200 run.

File: megatron/core/transformer/multi_token_prediction.py [hunk 9]
```diff
@@ -2407,8 +2461,22 @@ def custom_forward(
...
+                        shared_key, shared_topk_indices, shared_topk_length
+                    )
+                recompute_context = MTPDSAIterationContext(
+                    iteration=mtp_dsa_context.iteration, shared_tensors=shared_tensors
+                )
+
+            outputs = self._proj_and_transformer_layer(
                 hidden_states=hidden_states,
                 decoder_input=decoder_input,
                 input_ids=input_ids,
```
[the remark this answers is not in the document]
Comment (buptzyb) [reply megatron/core/transformer/multi_token_prediction.py:2484]: Kept intentionally and documented in `d8300ca38`. `topk_length` is first a checkpoint output, not only an input saved by `_save_args_to_ctx`; the source path needs a stable four-tensor output schema whether per-row lengths exist or not, and the same slot is passed positionally during consumer recomputation. The empty `int32` sentinel avoids backend-dependent `None` output/gradient handling while preserving that schema. Optional-length round trips are covered through native tensor-parallel checkpointing and the TE/MXFP8 recompute path, both passing on GB200.

File: docs/user-guide/features/multi_token_prediction.md [hunk 10]
```diff
@@ -55,4 +65,4 @@ Use `m` for MTP layers in the pipeline layout string. For example:
 
 ## Unsupported Combinations
 
-Context Parallel (CP), arbitrary `AttnMaskType`, and learned absolute position embeddings are not supported with MTP.
+Arbitrary `AttnMaskType` and learned absolute position embeddings are not supported with MTP. MTP supports Context Parallel execution for attention implementations that provide a compatible CP path; repeated DSA sharing requires DSA all-gather CP and reuses the iteration-0 global latent key within the same microbatch CP group.
```
[the remark this answers is not in the document]
Comment (buptzyb) [reply docs/user-guide/features/multi_token_prediction.md]: Fixed in `d8300ca38`. The documentation now states the supported path directly: repeated DSA sharing supports CP through `cp_comm_type=allgather` and reuses the iteration-0 global latent key inside the same microbatch CP group. Other attention implementations retain their existing MTP/CP behavior. The section also states that indexer loss is charged once and documents the speculative-decoding restriction. CP2 forward/backward coverage passed on GB200.
